pecos/core/xmc/inference.hpp

```diff
@@ -1125,6 +1126,101 @@ namespace pecos {
 
     }
 
+    // Prolongates the predictions of the previous layer to the select children of nodes.
+    // The result is returned as a csr_t matrix.
+    csr_t prolongate_sparse_predictions(const csr_t& csr_pred, const csc_t& C, const csr_t& csr_codes) {
+        typedef typename csr_t::mem_index_type mem_index_type;
+        typedef typename csr_t::index_type index_type;
+        typedef typename csr_t::value_type value_type;
+
+        auto rows = csr_codes.rows;
+        auto cols = csr_codes.cols;
+
+        // Compute the nnz's of each row
+        mem_index_type* row_ptr = new mem_index_type[rows + 1];
+        row_ptr[0] = 0;
+#pragma omp parallel for schedule(dynamic,4)
+        for (index_type row = 0; row < rows; ++row) {
+            index_type row_nnz = 0;
+
+            mem_index_type csr_codes_row_start = csr_codes.row_ptr[row];
+            mem_index_type csr_codes_row_end = csr_codes.row_ptr[row + 1];
+
+            std::unordered_set<index_type> valid_cols;
```

Comment:
Is it more efficient to pre-compute the col indices sets instead of constructing the set here since you are using those in both loops (Line1143 and Line1184)?

Reply:
Modified to pre-compute the valid col indices of each row. Re-verified correctness on Eurlex-4k and re-verified that there are no memory leaks.

Reply:
Changes have been undone. Upon further testing the speed of precomputing the indices is actually slower. Storing them also means it is less memory efficient.